src/main/java/org/apache/sysml/hops/codegen/opt/PlanSelectionFuseCostBasedV2.java

```diff
@@ -260,32 +263,40 @@ else if( USE_COST_PRUNING ) {
 				}
 			}
 			
-			//cost assignment on hops
-			double C = getPlanCost(memo, part, matPoints, plan, costs._computeCosts);
-			numEvalPlans ++;
-			if( LOG.isTraceEnabled() )
-				LOG.trace("Enum: "+Arrays.toString(plan)+" -> "+C);
-			
-			//cost comparisons
-			if( bestPlan == null || C < bestC ) {
-				bestC = C;
-				bestPlan = plan;
+			//cost assignment on hops. Stop early if exceeds bestC.
+			double C = getPlanCost(memo, part, matPoints, plan, costs._computeCosts, bestC);
+			if( C == Double.POSITIVE_INFINITY ) {
+				numEvalPartialPlans++;
 				if( LOG.isTraceEnabled() )
-					LOG.trace("Enum: Found new best plan.");
+					LOG.trace("Enum: stop costing early (exceed upper bound)");
+			} else {
+				numEvalPlans++;
+				if (LOG.isTraceEnabled())
+					LOG.trace("Enum: " + Arrays.toString(plan) + " -> " + C);
```

Comment:
Could you please merge these two branches? It would be good to retain the trace information of the current plan and its costs in both cases.